drivers/usb/udc/udc_ambiq.c

```diff
@@ -0,0 +1,925 @@
+
+	if (e_speed == AM_HAL_USB_SPEED_HIGH) {
+		return UDC_BUS_SPEED_HS;
+	} else {
+		return UDC_BUS_SPEED_FS;
+	}
+}
+
+static int udc_ambiq_ep_enqueue(const struct device *dev, struct udc_ep_config *ep_cfg,
+				struct net_buf *buf)
+{
+	struct udc_ambiq_event evt = {
+		.ep = ep_cfg->addr,
+		.type = UDC_AMBIQ_EVT_XFER,
+	};
+
+	LOG_DBG("%p enqueue %x %p", dev, ep_cfg->addr, buf);
+	udc_buf_put(ep_cfg, buf);
+	if (ep_cfg->addr == USB_CONTROL_EP_IN && buf->len == 0 && udc_ambiq_ctrl_pending_in_ack) {
+		udc_ambiq_ctrl_pending_in_ack = false;
+		udc_ep_xfer_complete_callback(USB_CONTROL_EP_IN, 0, 0, NULL);
+		return 0;
+	}
+	k_msgq_put(&drv_msgq, &evt, K_NO_WAIT);
+
+	return 0;
+}
+
+static int udc_ambiq_ep_dequeue(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	unsigned int lock_key;
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+	struct net_buf *buf;
+
+	lock_key = irq_lock();
+
+	buf = udc_buf_get_all(dev, ep_cfg->addr);
+	if (buf) {
+		udc_submit_ep_event(dev, buf, -ECONNABORTED);
+	}
+
+	udc_ep_set_busy(dev, ep_cfg->addr, false);
+	am_hal_usb_ep_state_reset(priv->usb_handle, ep_cfg->addr);
+	irq_unlock(lock_key);
+
+	LOG_DBG("dequeue ep 0x%02x", ep_cfg->addr);
+
+	return 0;
+}
+
+static int udc_ambiq_ep_set_halt(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	LOG_DBG("Halt ep 0x%02x", ep_cfg->addr);
+
+	am_hal_usb_ep_stall(priv->usb_handle, ep_cfg->addr);
+
+	return 0;
+}
+
+static int udc_ambiq_ep_clear_halt(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	LOG_DBG("Clear halt ep 0x%02x", ep_cfg->addr);
+
+	am_hal_usb_ep_clear_stall(priv->usb_handle, ep_cfg->addr);
+
+	return 0;
+}
+
+static int udc_ambiq_ep_enable(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+	uint8_t endpoint_type;
+	uint32_t status;
+
+	__ASSERT_NO_MSG(ep_cfg);
+
+	switch (ep_cfg->attributes) {
+	case USB_EP_TYPE_CONTROL:
+		endpoint_type = 0; /* AM_HAL_USB_EP_XFER_CONTROL */
+		break;
+	case USB_EP_TYPE_ISO:
+		endpoint_type = 1; /* AM_HAL_USB_EP_XFER_ISOCHRONOUS */
+		break;
+	case USB_EP_TYPE_BULK:
+		endpoint_type = 2; /* AM_HAL_USB_EP_XFER_BULK */
+		break;
+	case USB_EP_TYPE_INTERRUPT:
+		endpoint_type = 3; /* AM_HAL_USB_EP_XFER_INTERRUPT */
+		break;
+	default:
+		return -EINVAL;
+	}
+	status = am_hal_usb_ep_init(priv->usb_handle, ep_cfg->addr, endpoint_type, ep_cfg->mps);
+	if (status != AM_HAL_STATUS_SUCCESS) {
+		LOG_ERR("am_hal_usb_ep_init failed(0x%02x), %d", ep_cfg->addr, (int)status);
+		return -EIO;
+	}
+	LOG_DBG("Enable ep 0x%02x", ep_cfg->addr);
+
+	return 0;
+}
+
+static int udc_ambiq_ep_disable(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	__ASSERT_NO_MSG(ep_cfg);
+	am_hal_usb_ep_stall(priv->usb_handle, ep_cfg->addr);
```

Comment:
Why should disabled endpoint respond with stall handshake?

Reply:
Hi @jfischer-no, this is indeed a mistake. Thanks for catching this. It is corrected as follows: 
https://github.com/zephyrproject-rtos/zephyr/blob/ce491e0f5379c01cb5fb26bbbfe437e839b75a10/drivers/usb/udc/udc_ambiq.c#L353-L362